feat: created plugin for slowing down route or entire app (#2)
nearform/fastify-slow-down: README.md

![CI](https://github.com/nearform/bench-template/actions/workflows/ci.yml/badge.svg?event=push)

A slow down plugin for fastify

## Installation

```bash
npm i fastify-slow-down
```

> ## Usage

Reviewer: ```suggestion
## Usage

```

Register the plugin.

> This plugin will add an `onRequest` hook to slow down replies if a client (based on their IP address) has made too many multiple requests in the given `timeWindow`.

Reviewer: ```suggestion
This plugin will add an `onRequest` hook to slow down replies if a client (based on their IP address by default) has made too many multiple requests in the given `timeWindow`.
```

```js
import Fastify from 'fastify'
import slowDownPlugin from 'fastify-slow-down'

const fastify = Fastify()

// register the plugin
fastify.register(slowDownPlugin)

// create a route

> fastify.get('/', async (req, reply) => 'Hello fastify-slow-down!')

Reviewer: ```suggestion
fastify.get('/', async () => 'Hello from fastify-slow-down!')
```

// start server
await fastify.listen({ port: 3000 })
```

The response will have some additional headers:

| Header | Description |
|--------|-------------|
|`x-slow-down-limit`     | how many requests in total the client can make until the server starts to slow down the response
|`x-slow-down-remaining` | how many requests remain to the client in the `timeWindow`
|`x-slow-down-delay` | how much delay (in milliseconds) has been applied to this request

## Configuration

> - **delay**: Amount of delay imposed on request. it can be expressed in miliseconds or as string ([`ms`](https://github.com/zeit/ms) format). Set to `0` to disable delaying.

Reviewer: ```suggestion
- **delay**: Base unit of time delay applied to requests. It can be expressed in milliseconds or as string in [`ms`](https://github.com/zeit/ms) format. Set to `0` to disable delaying.
```

> - **timeWindow**: The duration of the time window. It can be expressed in milliseconds or as a string ([`ms`](https://github.com/zeit/ms) format). Set to `0` to disable delaying.

Reviewer: ```suggestion
- **timeWindow**: The duration of the time window during which request counts are kept in memory. It can be expressed in milliseconds or as a string in [`ms`](https://github.com/zeit/ms) format. Set to `0` to disable delaying.
```

> - **delayAfter**:  max number of connections during `timeWindow` before starting to delay responses with time specifed `delay`. Set to `0` to disable delaying.

Reviewer: ```suggestion
- **delayAfter**: number of requests received during `timeWindow` before starting to delay responses. Set to `0` to disable delaying.
```

> - **keyGenerator**: Function used to generate keys for saved records in memory.

Reviewer: format this as a table, with columns: name, type (in TS syntax), default value, description.

the description should explain what special values mean

> - **keyGenerator**: Function used to generate keys for saved records in memory.

Reviewer: ```suggestion
- **keyGenerator**: Function used to generate keys to uniquely identify requests coming from the same user
```

> ## Example with configuration

Reviewer: ```suggestion
## Example with configuration

```

> Giving the options it can be configurable for example: 

Reviewer: Registering the plugin with these options:

```js

> { 
>     delay: '10 seconds',
>      delayAfter: 10,
>      timeWindow: '10 minutes'
> }

Reviewer: ```suggestion
{ 
  delay: '10 seconds',
  delayAfter: 10,
  timeWindow: '10 minutes'
}
```

```

> The `delay` option will be applyied to the server when it is hitted for more then `delayAfter` option times in the specific `timeWindow` option

Reviewer: ```suggestion
A delay specified via the `delay` option will be applied to requests coming from the same IP address (by default) when more than `delayAfter` requests are received within the time specified in the `timeWindow` option.
```

***
In 10 minutes the result of hitting the API will be: 

1st request - no delay

2nd request - no delay

3rd request - no delay

```...```

10 request - no delay

11 request - 10 seconds  delay

12 request - 20 seconds  delay

13 request - 30 seconds  delay

```...```

20 request - 100 seconds  delay

```...```

***

> 

Reviewer: let's format this in a slightly better way, for example a unordered list, and explain that after 10 minutes, no delay at all is applied, meaning that the timeWindow is a fixed window, not a sliding window

```js
import Fastify from 'fastify'
import slowDownPlugin from 'fastify-slow-down'

const fastify = Fastify()

fastify.register(slowDownPlugin, { 
    delay: '10 seconds',
     delayAfter: 10,
     timeWindow: '10 minutes'
})

fastify.get('/', async (req, reply) => 'Hello fastify-slow-down!')

await fastify.listen({ port: 3000 })

> ```

Reviewer: remove this, we don't need to show another example
